### ISSUE TITLE
[BUG] No Colors for both `up_to_date` and `outdated`

### ISSUE THREAD

**Issue by Saplyn** [vuki656/package-info.nvim#155]
### Issues

- [X] I have checked [existing issues](https://github.com/vuki656/package-info.nvim/issues?q=is%3Aissue) and there are no issues with the same problem.

### Plugin Version

- [X] I am using the latest version of the plugin

### Neovim Version

- [X] I am using the 0.6 neovim version or later

### Neovim Version

NVIM v0.10.0 Build type: RelWithDebInfo LuaJIT 2.1.1713484068 Run "nvim -V1 -v" for more info

### Branch

master

### Actual behavior

Both `up_to_date` and `outdated` has no color, no matter the config.

> `vue` is "outdated"

![image](https://github.com/user-attachments/assets/35fe088f-8948-4899-b3e3-a6a1d635403b)

### Expected behavior

The color should appear, at least the default ones.

### Steps to reproduce

- Install the plugin
- Restart NeoVim
- Open a `package.json`
- Wait and observe the bug (no color)

### Package info config

I'm using [`lazy.nvim`](https://lazy.folke.io/) for package manager and [`LazyVim`](https://www.lazyvim.org/) distro.

```lua
-- lua/plugins/package-info.lua
return {
  "vuki656/package-info.nvim",
  dependencies = { "MunifTanjim/nui.nvim" },
  ft = "json",
  opts = {
    colors = {
      up_to_date = "#0DB9D7",
    },
    package_manager = "npm",
  },
}
```

### Other information

- I'm using gnome-terminal
- The output of `echo $TERM` is `xterm-256color`
- I'm using Ubuntu 22.04, with zsh & oh-my-zsh

### Help

No, sorry.

### Implementation help

Sorry I'm new to this editor, however I would try my best to help provide any information needed.

**Comment by GitMurf** [comment]
I am having the same issue. I am also using lazyvim. I wonder if it has to do with lazy loading?

**Comment by GitMurf** [comment]
I can confirm it seems the problem is when using any kind of lazy loading with lazyvim. If you set "lazy = false" it should work (but ofcourse is always loading at startup). @Saplyn can you confirm that lazy = false works for you as well?

**Comment by GitMurf** [comment]
I think it is related to #141

**Comment by GitMurf** [comment]
Ok I believe here is the problem: https://github.com/vuki656/package-info.nvim/blob/018bc10e28b00e28e239b1fff7c497304f8b1bf5/lua/package-info/config.lua#L123-L127

This runs during plugin setup and when not using lazyvim lazy loading then the color scheme will be "default" as it has not been set yet and therefore the `M.__register_highlight_groups()` function will run which registers the highlight groups.

When we use any sort of lazy loading then the color scheme is NOT default and therefore this does not run and the highlight groups never get registered.

I confirmed my hunch by NOT lazy loading which normally the highlight groups then get registered properly, but I added a vim.schedule(fn()) wrapper around the plugin setup() function so that it gets run "last" after the current call stack and by that time the problem exists as the color scheme is set and therefore no longer "default". Basically vim.schedule() is a way of representing the effect of lazy loading.

```lua
    config = function(_, opts)
        vim.schedule(function()
            require('package-info').setup(opts)
        end)
    end,
```

cc @vuki656

**Comment by Saplyn** [comment]
Thanks for you comments! The bad news is that your solution doesn't seem to work out for me, but good news is that I found a workaround:

```lua
return {
  "vuki656/package-info.nvim",
  dependencies = { "MunifTanjim/nui.nvim" },
  ft = "json",
  opts = {
    colors = {  -- specify both color
      up_to_date = "#0DB9D7",
      outdated = "#d19a66",
    },
  },
  config = function(_, opts)
    require("package-info").setup(opts)

    -- manually register them
    vim.cmd([[highlight PackageInfoUpToDateVersion guifg=]] .. opts.colors.up_to_date)
    vim.cmd([[highlight PackageInfoOutdatedVersion guifg=]] .. opts.colors.outdated)
  end,
}
```

This workaround works with or without lazy load, and demonstrates that the problem should be related to the highlight group does not correct registered.

**Comment by Saplyn** [comment]
I found that:

- My terminal works with `guifg=` but not `ctermfg=`, and `vim.o.termguicolors` is true, which means [this function should be correct](https://github.com/vuki656/package-info.nvim/blob/018bc10e28b00e28e239b1fff7c497304f8b1bf5/lua/package-info/utils/register-highlight-group.lua#L4C1-L13C4)
- When I comment out those two "manually register" lines, `:highlight PackageInfo` <kbd>tab</kbd> can't invoke any auto-complete, but when add those two lines back, <kbd>tab</kbd> auto-completes both groups. So the problem should be the hightlight group not registered.

**Comment by GitMurf** [comment]
Yes you are correct. The problem is highlight groups not being registered. The root cause of the problem is lazy loading because the plugin assumes you are loading the plugin immediately at startup. As described in more detail here: https://github.com/vuki656/package-info.nvim/issues/155#issuecomment-2270512637

I also added something similar where I registered the highlights in my config function which is a temporary fix for now. I also re-registered the autocmds the plugin adds to make sure they are done properly when lazy loading.

**Comment by vuki656** [comment]
Hey @GitMurf @Saplyn. Yeah I didn't take into account lazy loading when making the plugin.

Would any of you be willing to open a PR to fix this?

**Comment by folliehiyuki** [comment]
Is it better to allow the user to control the style/color by setting the highlight groups directly themselves (or inside their theme plugins), instead of using a configuration option?

When the plugin is loaded, it can check whether the highlight groups were already defined. If none is found, the default values will be set.

**Comment by vuki656** [comment]
> Is it better to allow the user to control the style/color by setting the highlight groups directly themselves (or inside their theme plugins), instead of using a configuration option?
> 
> When the plugin is loaded, it can check whether the highlight groups were already defined. If none is found, the default values will be set.

Yeah that makes more sense. Feel free to open a PR
